piilo/engines/anonymizer.py

```diff
@@ -95,71 +149,254 @@ def generate_surrogate(self, original_name: str) -> str:
+            surrogate = self.randomize_characters(pii)
+            self.seen_pii["USER_ID"][pii] = surrogate
+
+        surrogate = self.seen_pii["USER_ID"][pii]
+        self._operator_log(pii, surrogate)
+        return surrogate
+
+    def shuffle_location(self, pii: str) -> str:
+        return self.shuffle_obfuscate(pii, "LOCATION")
+
+    def shuffle_education(self, pii: str) -> str:
+        return self.shuffle_obfuscate(pii, "EDUCATION")
+
+    def shuffle_employer(self, pii: str) -> str:
+        return self.shuffle_obfuscate(pii, "EMPLOYER")
+
+    def shuffle_obfuscate(self, pii: str, pii_type: str) -> str:
+        """Return a previously seen obfuscated value for the provided pii_type"""
+
+        if pii not in self.seen_pii[pii_type]:
+            self.seen_pii[pii_type][pii] = random.choice(
+                self.all_pii.get(pii_type, [""])  # default to empty string
+            )
+
+        surrogate = self.seen_pii[pii_type][pii]
+        if pii != "PII":
+            logger.info(f"{pii} --> {surrogate}")
+        return surrogate
+
+    def map_obfuscate(self, pii: str) -> str:
+        """Return a surrogate value from the obfuscation map.
+        If no value is found, return an empty string.
+        If the value is "NOT_PII", return the input PII without obfuscation."""
+
+        surrogate = self.obfuscation_map.get(pii, "")
+
+        if surrogate == "NOT_PII":
+            surrogate = pii
+
+        if pii != "PII":
+            logger.info(f"{pii} --> {surrogate}")
+
+        return surrogate
+
+    def extract_ents(self, doc: Doc) -> list[RecognizerResult]:
+        """Return a list of RecognizerResults using the .ents attribute of
+        a Spacy Doc object."""
+
+        results = []
+        for ent in doc.ents:
+            spacy_result = RecognizerResult(
+                entity_type=ent.label_,
+                start=ent.start_char,
+                end=ent.end_char,
+                score=1.0,
+            )
+            results.append(spacy_result)
+
+        return results
+
+    def anonymize(self, text: str, analyzer_results: list[RecognizerResult]) -> str:
         """Anonymize identified input using Presidio Anonymizer."""
 
         if not text:
+            logging.warning("No input provided to anonymize.")
             return
 
-        analyzer_results = self._remove_conflicts_and_get_text_manipulation_data(
+        self.set_seeds(text)
+
+        if self.remember == "document":
+            # reset seen_pii for each document
+            self.seen_pii = defaultdict(dict)
+
+        entity_data = self._remove_conflicts_and_get_text_manipulation_data(
             analyzer_results, "merge_similar_or_contained"
         )
 
         operators = self._AnonymizerEngine__check_or_add_default_operator(
             {
                 "STUDENT": OperatorConfig(
-                    "custom", {"lambda": self.generate_surrogate}
+                    "custom", {"lambda": self.generate_surrogate_name}
```

Comment:
These don't seem to match the seven types used in the PIILO competition. Is the plan to use these instead?

Reply:
Ah, sorry, we can remove the methods for PII types that were not used in the competition. We obfuscated all PII in the competition dataset, but the competition only included labels for direct identifiers.

We don't have any methods for labeling these indirect identifiers currently.